### ISSUE TITLE
kubectl logs --follow shouldn't timeout or exit non-zero otherwise

### ISSUE THREAD

**Issue by dkarlovi** [kubernetes/kubectl#1373]
### What happened?

```
kubectl logs -f pod
```

with a pod silent for a while, say 10 mins, after a while it gets disconnected.

### What did you expect to happen?

The logs are streamed indefinitely. 

Alternatively, if a timeout is the cause of the interruption, exit with non-zero status.

### How can we reproduce it (as minimally and precisely as possible)?

Run a pod which does something like

```
date
sleep 3600
date
```

and watch its logs with `logs -f`

### Kubernetes version

<details>

```console
$ kubectl version
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.6", GitCommit:"ff2c119726cc1f8926fb0585c74b25921e866a28", GitTreeState:"archive", BuildDate:"2023-01-19T00:00:00Z", GoVersion:"go1.19.5", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v4.5.7
Server Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.5", GitCommit:"804d6167111f6858541cef440ccc53887fbbc96a", GitTreeState:"clean", BuildDate:"2022-12-11T00:17:11Z", GoVersion:"go1.19.4", Compiler:"gc", Platform:"linux/amd64"}

```

</details>

### Cloud provider

<details>
Azure AKS
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by mpuckett159** [comment]
/triage accepted
We would like to implement a non-zero exit code for a timeout/failure for sure, we will investigate the ability to ignore timeouts due to inactivity.

**Comment by XDRAGON2002** [comment]
Hey, would this be a good issue for a new contributor? If so, I would like to try and give it a shot.

**Comment by mpuckett159** [comment]
So I actually think that the root of the error here is that we have an `--ignore-errors` flag, but it defaults to true all the time. I haven't checked further than just skimming through the logs code really quick, however if we look [here](https://github.com/kubernetes/kubernetes/blob/d48b8167f7094bfae89ab601d2356868ba107627/staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go#L174) we see the definition of the flag itself, and I'm pretty sure it defaults to true, meaning that the only way to not ignore errors would be passing `--ignore-errors=false` I believe. Needs more investigation, but hopefully should be relatively straight forward if you want to give it a shot.

**Comment by XDRAGON2002** [comment]
@mpuckett159 sure, I'll give it a try, thanks!

/assign

**Comment by dkarlovi** [comment]
@mpuckett159 do you still think worthwhile to do some research if it's possible to workaround timeouts?

**Comment by XDRAGON2002** [comment]
Hey @mpuckett159 I had a small query, should I open the PR with the fix in this repo or in the main k8s one?

**Comment by mpuckett159** [comment]
Yes, all PRs for fixing code in kubectl should be opened in the main kubernetes/kubernetes repo, and code changes should be done in the staging directory. Also be sure to sign the CLA. You can find more details on contributing and your first contribution [here](https://www.kubernetes.dev/docs/guide/contributing/).

**Comment by XDRAGON2002** [comment]
Thanks @mpuckett159 for letting me know! Opened a PR fixing the same.

**Comment by mpuckett159** [comment]
Excellent I'll take a look.

**Comment by brianpursley** [comment]
By the way, I don't think `--ignore-errors` actually defaults to true, unless I am reading this wrong:

```
$ kubectl logs --help | grep -A1 ignore-errors
    --ignore-errors=false:
	If watching / following pod logs, allow for any errors that occur to be non-fatal

```

I'm pretty sure cobra uses whatever the value of the variable is, and `o.IgnoreLogErrors` appears to be an uninitialized bool when this runs:
```
	cmd.Flags().BoolVar(&o.IgnoreLogErrors, "ignore-errors", o.IgnoreLogErrors, "If watching / following pod logs, allow for any errors that occur to be non-fatal")
```

**Comment by brianpursley** [comment]
@dkarlovi.  Can you provide more details on this?

I am trying to reproduce and am unable.  I am using my own cluster and not Azure AKS, so maybe there is something provider-specific.  I am also using a newer client and server, so that could be a factor too.

Here is what I've tried:
```
$ kubectl run bb --image=busybox --command -- /bin/sh -c "date; sleep 3600; date"
```

```
$ kubectl logs -f bb
Fri May 12 23:20:56 UTC 2023
```

... and it runs for over 15 minutes, and I finally decided to try to make it exit, so I ran `kubectl edit pod bb` and changed the image to `busybox:1.35`, then my `kubectl logs` command exits with exit code 0.  It terminated normally because the pod stopped (restarted).

---

Now, if I force request timeout to something short, I get a timeout error and a nonzero exit code...

```
$ kubectl logs -f bb --request-timeout=5
Fri May 12 23:37:28 UTC 2023
error: net/http: request canceled (Client.Timeout or context cancellation while reading body)
$ echo $?
1
```

...but I don't think this is what this issue is referring to.

Any additional info would be welcome.

**Comment by dkarlovi** [comment]
@brianpursley maybe something in the release fixed it in the meantime, let me retry and update this issue, will get back to you here on Monday. Thanks!

**Comment by dkarlovi** [comment]
@brianpursley you're right, I can't reproduce it any longer either, it streamed the empty log for 1h correctly. Closing here, thanks for looking into this!

**Comment by dkarlovi** [comment]
BTW if anyone ends up finding this issue, `kubectl` 1.25.0 did have this issue for me, but 1.29 did not, everything else being equal. It was fixed at some point after 1.25.0 but before 1.29.0, I don't know how.
